fms_fsdp/policies/ac_handler.py

```diff
@@ -14,16 +14,33 @@
 )
 
 
-def apply_fsdp_checkpointing(model, every_xth_item):
+def apply_fsdp_checkpointing(model, p):
```

Comment:
can we write a test for this?

Reply:
think this way:

if we do cut_off = 1, this is essentialy:
1/p, 2/p, 3/p, 4/p, ...

And rounding to the nearest integers, we get what we want.

And minors are:
1. we add 1/2 to these 1,2,3,4... to make it more evenly spaced.    cut_off can 0 (left aligned), 1(right aligned) or 1/2 (central aligned)
2. we didn't do the rounding way directly as we might not know how many blocks in total beforehand. 


But really , it is basically doing 1/p, 2/p, 3/p, 4/p and round these numbers to integeters. so logically it is simple but code might be a little less straightforward to read

Reply:
right, that logic is a bit confusing to read. I think a small test could make it clearer in case we ever want to change things, and would serve as extra documentation too

Reply:
@nairbv added.

Reply:
why not add an actual test though? a test demonstrates and ensures specific examples work as expected, and gets run by CI so anything in the future that changes the behavior will get explicitly called out. An automated test can also show a lot of examples quickly.

Even just to write out the example like `p = 1/3: [no-ac, ac, no-ac, no-ac, ac, no-ac,` ... to be sure it's correct you must have tested it somehow? if a user wants to reproduce that, it's nice to be able to see how you got the results, so helpful to paste whatever you wrote into a pytest. here's what I've got locally:

```
import torch
from fms.models.llama import LLaMABlock, LLaMAConfig
from fms.modules.positions import RotaryEmbedding

def checkpointing_selector(p):
    block_idx=0
    cut_off = 1/2
    def selective_checkpointing(submodule):
        nonlocal block_idx
        nonlocal cut_off
        if isinstance(submodule, LLaMABlock):
            block_idx += 1
            if block_idx * p >= cut_off:
                cut_off += 1
                return True
            return False
    return selective_checkpointing

config = LLaMAConfig(src_vocab_size=10, emb_dim=10, nheads=1)
rot = RotaryEmbedding(dim=10, max_seq_len=10)
selector = checkpointing_selector(0.25)
modules = [LLaMABlock(config, rot) for _ in range(10)]
selected = [selector(block) for block in modules]

expected = [False, True, False, False, False, True, False, False, False, True]

print(selected, expected)
assert selected == expected
```

Reply:
Got it. I was initially hesitated to add a test case for this because essentially we are only testing a math equation of `var a * const b >=  const c` and I wasn't sure if this is worth testing.

But I get your point, simple doesn't mean obvious. And it might be over-obvious to me only because I wrote this.

I will add one when I get some time.

Reply:
I think the following should be enough to get what you had above:

```python
model = LLaMA(LLaMAConfig(emb_dim=16, nheads=1,nlayers=10))
apply_fsdp_checkpointing(model, "1/3")
print([isinstance(block, CheckpointWrapper) for block in model.layers])
```

And I cam parameterize it to test various of p values.